### PR TITLE
[FVM] Basic fuzzing and minor test change

### PR DESCRIPTION
**Pull request by janezpodhostnik**
Add some basic fuzzing to the FVM. 

For now the only way to run this is manually.

I had it running for a while, and nothing was found :tada:.

### CODE REVIEW

File: fvm/fvm_test.go [hunk 1]
```diff
@@ -85,6 +95,60 @@ func (vmt vmTest) run(
 	}
 }
 
+// bootstrapWith executes the bootstrap procedure and the custom bootstrap function
```
Comment (pattyshack) [inline fvm/fvm_test.go:98]: maybe move this to fuzz test if you're not going to reuse it elsewhere

File: fvm/fvm_fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+)
+
+func FuzzTransactionComputationLimit(f *testing.F) {
+	// setup initial state
+	vmt, tctx := bootstrapFuzzStateAndTxContext(f)
+
+	f.Add(uint64(0), uint64(0), uint(0))
+	f.Add(uint64(5), uint64(0), uint(0))
+	f.Fuzz(func(t *testing.T, computationLimit uint64, memoryLimit uint64, transactionType uint) {
+		computationLimit %= flow.DefaultMaxTransactionGasLimit
```
Comment (pattyshack) [inline fvm/fvm_fuzz_test.go:30]: why %=?  should be min?

File: fvm/fvm_fuzz_test.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		// ==== Create an account ====
+		var txBody *flow.TransactionBody
+		privateKey, txBody = testutil.CreateAccountCreationTransaction(tb, chain)
+
+		err := testutil.SignTransactionAsServiceAccount(txBody, 0, chain)
+		if err != nil {
+			return err
+		}
+
+		tx := fvm.Transaction(txBody, 0)
```
Comment (pattyshack) [inline fvm/fvm_fuzz_test.go:212]: can you use incrementing tx index?  you can use `programs.NextTxIndexForTesting()` if you don't want to keep track of the tx index manually (same elsewhere)

**Review by pattyshack — APPROVED** [review]

File: fvm/fvm_test.go (hunk 1 above)
Comment (janezpodhostnik) [reply fvm/fvm_test.go:98]: Using this, the fvm tests, especially those that repeat the same setup, could be much faster. I think I will be using this in the future.

File: fvm/fvm_fuzz_test.go (hunk 2 above)
Comment (bluesign) [reply fvm/fvm_fuzz_test.go:30]: it is something like: `random(0, flow.DefaultMaxTransactionGasLimit)`

Comment (janezpodhostnik) [reply fvm/fvm_fuzz_test.go:30]: I did some searching.

I do want randomness here. that is the whole point of the fuzzing tests. `%` preserves the randomness of whatever the fuzzing framework provides. Another option would be skipping if out of bounds. not sure I like that.

File: fvm/fvm_fuzz_test.go (hunk 3 above)
Comment (Tonix517) [reply fvm/fvm_fuzz_test.go]: +1

**Review by Tonix517 — APPROVED** [review]

**Comment by janezpodhostnik** [comment]
bors merge

**Review by zhangchiqing — APPROVED** [review]

File: fvm/fvm_fuzz_test.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+type transactionType struct {
+	createTxBody func(t *testing.T, tctx transactionTypeContext) *flow.TransactionBody
+	require      func(t *testing.T, tctx transactionTypeContext, results fuzzResults)
+}
+
+var fuzzTransactionTypes = []transactionType{
+	{
+		createTxBody: func(t *testing.T, tctx transactionTypeContext) *flow.TransactionBody {
```
[the remark this answers is not in the document]
Comment (janezpodhostnik) [reply fvm/fvm_fuzz_test.go:85]: This does more then just verify fee deduction, It also verifies there is no fatal errors and that only the expected error codes are returned. But making it clearer which transaction type we are using in which example is a good idea.

**Comment by janezpodhostnik** [comment]
bors merge

**Pull request merged**